core-primitives/stf-state-handler/src/state_snapshot_repository.rs

```diff
@@ -0,0 +1,447 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::{
+	error::{Error, Result},
+	file_io::StateFileIo,
+	state_snapshot_primitives::{
+		generate_current_timestamp_file_name, initialize_shard_with_file, SnapshotHistory,
+		StateFileMetaData,
+	},
+};
+use core::ops::RangeBounds;
+use itp_types::ShardIdentifier;
+use log::*;
+use std::{
+	collections::VecDeque, fmt::Debug, format, marker::PhantomData, string::String, sync::Arc,
+	vec::Vec,
+};
+
+/// Trait for versioned state access. Manages history of state snapshots.
+pub trait VersionedStateAccess {
+	type StateType;
+	type HashType;
+
+	/// Load the latest version of the state.
+	fn load_latest(&self, shard_identifier: &ShardIdentifier) -> Result<Self::StateType>;
+
+	/// Update the state, returning the hash of the state.
+	fn update(
+		&mut self,
+		shard_identifier: &ShardIdentifier,
+		state: Self::StateType,
+	) -> Result<Self::HashType>;
+
+	/// Reverts the current state to a version defined by a given state hash.
+	fn revert_to(
+		&mut self,
+		shard_identifier: &ShardIdentifier,
+		state_hash: &Self::HashType,
+	) -> Result<Self::StateType>;
+
+	/// Initialize a new shard.
+	fn initialize_new_shard(
+		&mut self,
+		shard_identifier: &ShardIdentifier,
+	) -> Result<Self::HashType>;
+
+	/// Checks if a shard for a given identifier exists.
+	fn shard_exists(&self, shard_identifier: &ShardIdentifier) -> bool;
+
+	/// Lists all shards.
+	fn list_shards(&self) -> Result<Vec<ShardIdentifier>>;
+}
+
+/// State snapshot repository.
```

Comment:
Why does this level need to know about repository or files?  Wouldn't it be enough if it knows about the `shard` which may have multiple `snapshots`? The StateFileIo or a layer inbetween should be able to draw the connection of shard and directory. Sorry, I keep going on about this, but I think we should draw a clean line here.

Reply:
The snapshot repository (in its current form) manages all the files itself. There is no abstraction layer in between the snapshot repository and shard directories with state files. The snapshot repository creates and deletes files directly.

What you're suggesting would certainly make sense, but is out of scope for this issue/PR. It will be addressed with the whole in-memory caching.

Reply:
Yes, this is changed now with the newest update 👍